database/migrations/2022_02_02_203008_filesize_size_variants.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Schema;
+
+/**
+ * Add a filesize column for size variants to make it
+ * easier to implement filesystems with a Lychee backend
+ * that needs to know in advance the size of files, for example.
+ *
+ * The size of the original photo is also moved to the original size variant.
+ */
+class FilesizeSizeVariants extends Migration
+{
+	private const VAR_TAB = 'size_variants';
+	private const PHOTO_FK = 'photo_id';
+	private const TYPE_COL = 'type';
+	private const WIDTH_COL = 'width';
+	private const HEIGHT_COL = 'height';
+	private const ID_COL = 'id';
+	private const TYPE_ORIGINAL = 0;
+	private const PHOTOS_TAB = 'photos';
+	private const SIZE_COL = 'filesize';
+
+	/**
+	 * Run the migrations.
+	 *
+	 * @return void
+	 */
+	public function up()
+	{
+		// To avoid doing I/O on every photo, which would be prohibitive
+		// on large instances, and because JPEG compression makes approximations
+		// from original filesize too approximative, just set filesize to -1
+		// for every variant except the original, indicating that the size
+		// has not been calculated yes.
+		// The real calculation can be done by calling `artisan lychee:variant_filesize` from the CLI.
+		// This will replace the default value by the actual filesize but takes time.
+		Schema::table(self::VAR_TAB, function (Blueprint $table) {
+			$table->bigInteger(self::SIZE_COL)->nullable(false)->default(-1);
```

Comment:
Have you checked if the foreign key constraint from `size_variants.photo_id` to `photo.id` remains intact if you alter the table scheme (i.e. add a column here and drop a column during the reverse migration)?

I had some bad experiences with that in the past. Eloquent tends to drop foreign keys and column restrictions of the whole table, if you modify the table scheme even if the modified/added/dropped column is completely unrelated to the constraints. Even if it does not happen for MySQL or PostgreSQL it sometimes happens with SQLite.

Reply:
Hmmm, I remember that during my test, one time, SQLite complained about a foreign key drop, but it never happened again.

How can I make sure, in the code, not to drop restrictions of the whole table ?

Reply:
I don't know any better way than testing with every DB engine and manually comparing the table definitions before and after the migration.

For PostgreSQL it is `\d size_variants`, for SQLite it is `.schema size_variants` and for MySQL (I assume) `SHOW CREATE TABLE size_variants`. In the case of MySQL, I am not sure if this command also shows the foreign key constraints or if you need another command.

Reply:
I managed to reproduce the problem I've had.

When trying to migrate existing data with SQLite, I got :

```
SQLSTATE[23000]: Integrity constraint violation: 19 FOREIGN KEY constraint failed (SQL: DROP TABLE photos)
```

with error pointing [from there](https://github.com/LycheeOrg/Lychee/pull/1239/files#diff-9a22abf7cfc5b5a793e236cd8d4a040d80f66d308a0457ca3617a5b4fa0dcafcR66) - i.e. the `dropColumn`. How could a  `dropColumn` ending up trying to delete the `photos` table ??

Reply:
So my gut feeling was right.

Dropping a column in SQLite is implemented by renaming the existing table to some temporary name (e.g. `size_variants_old`), re-creating the table with the one column removed, copying all data from the renamed, old table to the freshly created one and finally deleting the whole old old table.

Sadly, this implementation does not take foreign constraints into account.

There are only two options here:

a) Do the same as above manually, but also properly handle foreign constraints. Unfortunately, this will quickly evolve into a mess, because eventually you will end up re-creating the whole database (size variants will lead to photos, photos to albums, photos and albums to users, ...)

b) Do not drop the column on the downward migration.

Reply:
@nagmat84 The problem happens in the upward migration, when trying to drop the `filesize` column of the `photos` table.

Reply:
Also as indicated [there](https://www.sqlite.org/lang_altertable.html) the `filesize` column seems to check all conditions to that SQLite manages to drop the column.

Even doing it manually :
```
alter table size_variants add column filesize integer not null default '0';
alter table photos drop column filesize;
```

throws no error. I'll try to investigate further...

Reply:
I didn't see that the upward migration requires a `DROP COLUMN`, too. The explanation remains valid though, i.e. dropping a column from `photos` requires to rename the table, etc.

That's bad, because we obviously need to upward migration.

I checked with the SQLite documentation. It seems that SQLite should support `DROP COLUMN`, so you could try a `DB::raw` statement.

Reply:
Oh, so you mean that Eloquent is implementing column drop with SQLite this way ? Seems a bit odd. I should probably drop the column with a raw statement, thanks for your time and advice.

Reply:
I used a raw statement to delete the column in e350d0db7bb8b8785fe354991a3944d4f3b3cb2a.

Reply:
Looks good from inspection. I cannot test it myself right now, because I am on a business trip and I don't have a development environment at hand. But I assume you tested it.

From the look of it, it seems good. :+1: Conversation can be resolved.

Reply:
@nagmat84 Indeed, the ubuntu image uses [SQLite v3.31.1](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md) whereas `DROP COLUMN` is supported from [v3.35.0](https://www.sqlite.org/releaselog/3_35_0.html) (dec. 2021)

This version is quite new and not included yet in LTS distributions repositories... thus is seems hard to ask users to have that minimum version, right ? :/

Reply:
> This version is quite new and not included yet in LTS distributions repositories... thus is seems hard to ask users to have that minimum version, right ? :/

I see you haven't met our PHP requirements :laughing:

Reply:
Haha, indeed PHP 8 is not in bullseye either but in bookworm, as SQLite 3.35. But I feel that requiring this version only for a feature that nobody will use is maybe a bit harsh. What would you do ? Try to migrate the table with copy/drop "by hand" ?

Reply:
A migration "by hand" is not as easy as you might think, because dropping the tables `photos` and re-creating it is like opening a can of worms. You will end up re-creating nearly the whole DB due to the foreign constraints. If you want to see what happens, I recommend you to look at https://github.com/LycheeOrg/Lychee/blob/a8c1354890fdd8d49b3a73931764df1b5ed566b6/database/migrations/2021_12_04_181200_refactor_models.php It is a tiny, small, innocent migration of only 2000 LOC. 😂

I am not sure how many users actually use SQLite. Unfortunately, we don't have any usage statistics about that. However, it is not the recommended solution. Maybe we could require a higher minimal version for SQLite without annoying too many users.

I am unsure. @ildyria should decide.

Reply:
I guess the other option is to leave the column there and null and try to remember to drop it later, when it's more widely available? That might not help with the `down` though. And I'm not sure how many other things that would mess up.

Reply:
I propose the following compromise and like all compromises it is really ugly, horrible and hack-ish. I am ashamed of myself for even suggesting it.

If the migration detects MySQL or PostgreSQL as the DB engine, it could execute the raw DB statement to drop the column `filesize`. If the migration detects SQLite, it simply does nothing but leaves the column as it. The column won't be used anymore, but simply remains as it is. I don't like the idea of having an obsolete, unmaintained, abandoned column in the DB, because everybody will wonder what it is for in the future, but it is the least worst solution I can think of.

Note, the column has a set default value, so it does not interfere with future `INSERT` statements which don't provide a value for that column. On top, we could also update all existing values to `0` to indicate that the column is meaningless.

Reply:
> I guess the other option is to leave the column there and null and try to remember to drop it later, when it's more widely available? That might not help with the `down` though. And I'm not sure how many other things that would mess up.

You were faster than me with the same ugly idea. :laughing:

Reply:
Yeah, I'd leave it consistent across engines though. Then maybe open an issue against a future milestone (5?) so that we're somewhat less likely to forget.

Reply:
What is the matter of state for this aspect? I assume that @d7415 and I have already agreed to not drop the column now, but leave it untouched and open an issue such that we remember to drop the column with the next big re-factoring (although I don't know if and when this will happen) or after SQLite 3.35 has become more widely spread.

What is the opinion of the other developers, esp. @ildyria? Is this solution acceptable?

I would really like to see this PR being merged soon.

Reply:
> I would really like to see this PR being merged soon.

Right, I'll send a commit tonight or tomorrow!

Reply:
See cc9d88f905ec5a13157a19b6fb875d09f6ab2803.

I don't think it is worth trying to set the column nullable and I just changed the "old" column values so they are all zeros.

I also added a check in the upward migrations to avoid creating the column which already would exist after a cycle of migrate/rollback.

I tested uploading an image before the migration, migrate, launch command, rollback, migrate again (while switching code) and it seems good to me. Hope it is... :p

Reply:
I don't get why tests are failing with bad JSON (as if no size variants are created) while they do work in my environment. What could I do?

Reply:
> I don't think it is worth trying to set the column nullable and I just changed the "old" column values so they are all zeros.

Totally agreed. You must not even try to make the column nullable, because you would run into the same problem you have with dropping a column. Making a non-nullable column nullable in SQLite means to create a new column (with the same definition minus nullability), copying over all values, dropping the old column and renaming the new column to the old one. Hence, setting everything to `0` is just fine.

> I also added a check in the upward migrations to avoid creating the column which already would exist after a cycle of migrate/rollback.

:+1:

> I tested uploading an image before the migration, migrate, launch command, rollback, migrate again (while switching code) and it seems good to me. Hope it is... :p

Seems so.

Reply:
> I don't get why tests are failing with bad JSON (as if no size variants are created) while they do work in my environment. What could I do?

That is really weird, in particular as the tests succeed with SQLite as well as PostgreSQL and the choice of the DB engine should not change the result of this test at all.

Did you run the test with MySQL locally, too?

I could try to checkout your branch to my machine and see if I find something in the step debugger. But this might take some time.

Reply:
Well, again, my test environment was bad because I only tested SQLite. Now, I know I should test PostgreSQL, MariaDB & SQLite.

Problem was I forgot to remove the temporary `-1` filesize when creating variants. SQLite does not seem to enforce that filesize is signed, but MariaDB does. See 6c241babc7f8d6915cf95c73b2d72236ce4a03ae.

Maybe now I'm ready for a future PR with less bugs. :smile:

Reply:
I’m glad you found the solution. :tada: :confetti_balls: Congratulations!

I would say this aspect can be regarded as resolved now.